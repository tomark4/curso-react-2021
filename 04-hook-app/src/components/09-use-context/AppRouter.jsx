import {
    BrowserRouter as Router,
    Switch,
    Route
} from 'react-router-dom';

import AboutScreen from "./AboutScreen";
import HomeScreen from './HomeScreen';
import LoginScreen from './LoginScreen';
import NavBar from './NavBar';
import NotFoundPage from './NotFoundScreen';

const AppRouter = () => {
    return (
        <Router>
            <div>
                <NavBar />
                <Switch>
                    <Route path="/" component={HomeScreen} exact></Route>
                    <Route path="/login" component={LoginScreen} exact></Route>
                    <Route path="/about" component={AboutScreen} exact></Route>
                    <Route component={ NotFoundPage } exact></Route>
                </Switch>
            </div>
        </Router>
    )
}

export default AppRouter
