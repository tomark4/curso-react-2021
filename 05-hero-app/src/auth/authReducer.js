import { types } from '../types/types'

const authReducer = (state = {}, action) => {
  // reducers para manejar el login
  switch(action.type){
    case types.login:
      return {...action.payload, logged: true}

    case types.logout:
      return { logged: false }

    default:
      return state
  }
}

export default authReducer;
