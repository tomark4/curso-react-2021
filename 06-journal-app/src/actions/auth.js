import { types } from '../types/types';
import {firebase, googleAuthProvider} from '../firebase';
import { startLoading, stopLoading } from './ui';
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

// login with email and password
export const loginWithEmailPassword = (email, password) => (dispatch) => {
  dispatch( startLoading() );
  firebase.auth().signInWithEmailAndPassword(email, password)
  .then( ({user}) =>  {
    dispatch({
      type: types.login,
      payload: { uid: user.uid, name: user.displayName }
    });
    dispatch( stopLoading() );
  }).catch(err => {
    console.log(err);
    Swal.fire('Error',err.message, 'error')
    dispatch( stopLoading() );
  });
}

// login con google
export const loginWithGoogle = () => (dispatch) => {
  firebase.auth().signInWithPopup( googleAuthProvider )
  .then( ({user}) => {
    dispatch({
      type: types.login,
      payload: {uid: user.uid, name: user.displayName}
    })
  })
}

// register con email and password
export const registerWithEmailPassword = (email, password, name) => (dispatch) => {
  firebase.auth().createUserWithEmailAndPassword(email, password)
  .then( async ({user}) => {
    await user.updateProfile({displayName: name});
    dispatch({ type: types.login, payload: {uid: user.uid, name: user.displayName} });
  }).catch( err => console.log(err) );
}


export const startLogout = () => async (dispatch) => {
  try{
    await firebase.auth().signOut();
    dispatch( noteLogout() );
    dispatch( logout() );
  }catch(err){
    console.log(err)
  }
}

export const logout = () => ({
  type: types.logout
})
