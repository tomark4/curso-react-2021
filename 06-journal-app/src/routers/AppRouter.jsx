import {
    BrowserRouter as Router ,
    Route,
    Switch,
    Redirect
} from 'react-router-dom';
import AuthRouter from './AuthRouter';
import JournalScreen from '../components/journal/JournalScreen';

const AppRouter = () => {
    return (
        <Router>
            <div>
                <Switch>
                    <Route path="/auth" component={AuthRouter}/>
                    <Route path="/" component={JournalScreen} exact/>
                    <Redirect to="/auth/login" />
                </Switch>
            </div>
        </Router>
    )
}

export default AppRouter
