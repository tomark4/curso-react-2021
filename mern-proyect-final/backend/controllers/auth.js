const { response } = require('express');
const Usuario = require("../models/Usuario");
const bcrypt = require("bcryptjs");
const { generateJWT } = require("../helpers/jwt");

const register = async (req, res = response) => {
  try{
    const { name, email, password } = req.body;

    // validar que el email no exista
    let usuario = await Usuario.findOne({email});

    if(usuario){
      return res.status(400).json({
        ok: false,
        message: 'usuario existe con ese correo'
      });
    }

    // crear si no existe
    usuario = new Usuario({name, email, password});

    // encriptar la constraseña
    const salt = bcrypt.genSaltSync();
    usuario.password = bcrypt.hashSync( password, salt);
    await usuario.save();

    const token = await generateJWT(usuario.id, usuario.name);

    return res.status(201).json({
      ok: true,
      uid: usuario.id,
      name: usuario.name,
      message: 'user created',
      token
    });

  } catch(err){
    console.log(err);
    return res.status(500).json({
      ok: false,
      message: 'error, contact support'
    });
  }

}


const login = async (req, res = response) => {
  try{
    const { email, password } = req.body;
    const usuario = await Usuario.findOne({email});

    if(!usuario){
      return res.status(400).json({
        ok: false,
        message: 'Usuario y contraseña no son correctos 1'
      });
    }

    // confirmar los password
    const validPassword = bcrypt.compareSync(password, usuario.password);

    if(!validPassword){
      return res.status(400).json({
        ok: false,
        message: 'Usuario y contraseña no son correctos 2'
      });
    }

    const token = await generateJWT(usuario.id, usuario.name);

    return res.status(200).json({
      ok: true,
      uid: usuario.id,
      name: usuario.name,
      message: 'login success',
      token
    });

  } catch(err){
    console.log(err);
    return res.status(500).json({
      ok: false,
      message: 'error contact support'
    });
  }
}


const refreshToken = async (req, res = response) => {
  try{

    const { uid, name } = req;
    const token = await generateJWT(uid, name);
    return res.json({ok: true, message:"refresh token", token});

  } catch( err ){
    console.log(err);
    return res.status(500).json({ok: false, message:"error"});
  }
}


module.exports = {
  register,
  login,
  refreshToken
};
