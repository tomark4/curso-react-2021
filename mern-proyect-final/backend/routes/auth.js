const express = require('express');
const router = express.Router();
const controller = require('../controllers/auth');
const {check} = require("express-validator");
const validateFields = require('../middlewares/validate-fields');
const { validateJWT } = require("../middlewares/validate-jwt");

// register user
router.post("/register",[
  check('name','Name is required').not().isEmpty(),
  check('email','Email is required').isEmail(),
  check('password','Password must have 6 characters').isLength({min: 6}),
  validateFields
], controller.register);

// login user
router.post("/login", [
  check('email','Email is required').isEmail(),
  check('password','Password must have 6 characters').isLength({min: 6}),
  validateFields
], controller.login);

// refresh token
router.get("/refresh-token", [validateJWT], controller.refreshToken);


module.exports = router;
