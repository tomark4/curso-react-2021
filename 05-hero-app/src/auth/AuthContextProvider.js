import { createContext, useReducer, useEffect } from 'react'
import authReducer from './authReducer'

export const AuthContext = createContext()

const init = () => {
  return JSON.parse(localStorage.getItem('user')) || {logged: false};
}

const AuthContextProvider = ({children}) => {
  const [user, dispatch] = useReducer(authReducer, {}, init)

  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user));
  },[user])

  return (
    <AuthContext.Provider value={{user, dispatch}}>
      {children}
    </AuthContext.Provider>
  )
}

export default AuthContextProvider;
