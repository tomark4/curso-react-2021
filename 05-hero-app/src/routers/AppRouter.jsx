import { useContext } from 'react';
import {
    BrowserRouter as Router,
    Switch
} from 'react-router-dom';
import { AuthContext } from '../auth/AuthContextProvider';

import LoginScreen from '../components/screens/LoginScreen';
import DashboardRoutes from './DashboardRoutes';
import PrivateRoute from './PrivateRoute';
import PublicRoute from './PublicRoute';

const AppRouter = () => {

    const { user:{logged} } = useContext(AuthContext);

    return (
        <Router>
            <div>
                <Switch>
                    <PublicRoute exact
                    isAuth={logged}
                    path='/login'
                    component={LoginScreen} />

                    <PrivateRoute
                    path='/'
                    component={DashboardRoutes}
                    isAuth={logged}
                    />
                </Switch>
            </div>
        </Router>
    )
}

export default AppRouter
