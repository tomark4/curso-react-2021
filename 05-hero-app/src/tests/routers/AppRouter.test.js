import {mount} from 'enzyme'
import AppRouter from '../../routers/AppRouter'
import {AuthContext} from '../../auth/AuthContextProvider';

describe('prueba en el componente authrouter', () => {

  test('debe mostrar el login si no esta autenticado', () => {
    const wrapper = mount(
      <AuthContext.Provider value={{
        dispatch: jest.fn(),
        user:{
          logged: false
        }
      }} >
        <AppRouter />
      </AuthContext.Provider>
    );

    expect(wrapper).toMatchSnapshot()
  });

  test('debe de mostrar el componente marvel si esta autenticado', () => {
    const wrapper = mount(
      <AuthContext.Provider value={{
        dispatch: jest.fn(),
        user:{
          name: 'jose',
          logged: true
        }
      }} >
        <AppRouter />
      </AuthContext.Provider>
    );

    expect( wrapper.find('.navbar').exists() ).toBe(true);

  })


});
