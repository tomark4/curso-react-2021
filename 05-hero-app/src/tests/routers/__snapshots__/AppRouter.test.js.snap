// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prueba en el componente authrouter debe mostrar el login si no esta autenticado 1`] = `
<div>
  <div>
    <div
      className="container mt-5"
    >
      <div
        className="row"
      >
        <div
          className="col-6 mx-auto"
        >
          <div
            className="card"
          >
            <div
              className="card-body"
            >
              <h4>
                Login
              </h4>
              <hr />
              <button
                className="btn btn-success btn-block"
                onClick={[Function]}
              >
                Login
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
